src/server/api/general.js

```diff
@@ -166,17 +179,38 @@ export default function useGeneralApi(app) {
                 }
             }
 
-            yield createAccount({
-                signingKey: config.get('registrar.signing_key'),
-                fee: config.get('registrar.fee'),
-                creator: config.get('registrar.account'),
-                new_account_name: account.name,
-                delegation: config.get('registrar.delegation'),
-                owner: account.owner_key,
-                active: account.active_key,
-                posting: account.posting_key,
-                memo: account.memo_key,
-            });
+            // Ensure another registration is not in progress.
+            // Raw query with SQL_NO_CACHE avoids the MySQL query cache.
+            const newCreationHash = yield models.sequelize.query(
+                'SELECT SQL_NO_CACHE creation_hash FROM users WHERE id = ?',
+                {
+                    replacements: [user.id],
+                    type: models.sequelize.QueryTypes.SELECT,
+                }
+            );
+
+            if (newCreationHash[0].creation_hash !== creationHash) {
+                console.log({ newCreationHash, creationHash });
+                throw new Error('Creation hash mismatch');
+            }
+
+            try {
+                yield createAccount({
+                    signingKey: config.get('registrar.signing_key'),
+                    fee: config.get('registrar.fee'),
+                    creator: config.get('registrar.account'),
+                    new_account_name: account.name,
+                    delegation: config.get('registrar.delegation'),
+                    owner: account.owner_key,
+                    active: account.active_key,
+                    posting: account.posting_key,
+                    memo: account.memo_key,
+                });
+            } catch (e) {
+                yield user.update({ creation_hash: null });
+                throw new Error('Account creation error, try again.');
```

Comment:
when will this execute? I'm not 100% familiar with JS generator syntax, but in Python, execution would stop right here and you wouldn't reach the `throw` until the next invocation of the method

Reply:
this generator method is getting called over and over again, so the `throw` happens after the call to `user` is done -- doing the `yield` just lets Koa will move on to other work if the sql takes a minute